Add active low support

mcampbellizo opened the pull request:
Add the ability to setup GPIOs as active_low.

vitiral commented on gpio.py:
```diff
@@ -17,9 +17,10 @@ class PinState(object):
         direction: the file pointer to set/read direction of the pin.
     """
-    def __init__(self, value, direction):
+    def __init__(self, value, direction, active_low):
```
should have a default. `active_low = None`. If it is `None` it should not pass it to the `PinState`

mcampbellizo replied:
Due to the wav `_verify()` woks, it's never possible for `active_low` to be None. Not sure exactly what you mean by not passing to `PinState` if it's `None`, as it's already passed here. I changed `setup()` to allow `active_low` to be `None` at that point which I think address what you are getting at. Let me know if that is not true.

vitiral commented on gpio.py:
```diff
@@ -102,25 +104,36 @@ def cleanup(pin=None, assert_exists=False):
 
 @_verify
-def setup(pin, mode, pullup=None, initial=False):
+def setup(pin, mode, pullup=None, initial=False, active_low=False):
```
`None` and `None` needs to be a valid value.

mcampbellizo replied:
Not sure if this was a typo when you said `None` twice here. I changed the default value to be `None` and added logic that will preserve the original behavior if that is the value. This should preserve API and functional backwards compatibility.

vitiral replied:
I have no idea what I meant, haha -- but what you did sounds like what I would mean if I had been meaning something remotely intelligent, as opposed to... what I said... :wink:

vitiral commented on gpio.py:
```diff
@@ -17,9 +17,10 @@ class PinState(object):
         value: the file pointer to set/read value of pin.
         direction: the file pointer to set/read direction of the pin.
```
Add docs for active_low, including default behavior of `None` value

mcampbellizo replied:
See my comment below, I don't believe this value can be `None`, therefore it doesn't need docs for that. I added docs about the new `None` value of `active_low` to `setup()`

vitiral replied:
You're right, I didn't fully understand the code path.

mcampbellizo commented:
First, thanks for the quick review and feedback.

Totally agree with the sentiment of full API and functional backwards compatibility. To address this I added a default value for `active_low` in `setup()` of `None`. In this case it will not touch the active_low file which is the same behavior as before and requires no code changes on the consumer side to get this.

The only slight difference of under the hood behavior is that the `PinState` object will open the active_low file. This is unavoidable without a much larger refactor as the `_verify()` decorator can't know if `active_low` was passed is as it is obfuscated through either `*args` or `**kwargs`

I've added responses to the inline review above. Let me know if there is anything else.

vitiral commented on gpio.py:
```diff
@@ -102,25 +105,38 @@ def cleanup(pin=None, assert_exists=False):
     if mode not in (IN, OUT, LOW, HIGH):
         raise ValueError(mode)
 
+    if not type(active_low) == bool:
```
`if not isinstance(active_low, bool):`

And move into the `if active_low is not None` block below.

mcampbellizo replied:
sounds good.

vitiral commented on gpio.py:
```diff
@@ -102,25 +105,38 @@ def cleanup(pin=None, assert_exists=False):
+    if active_low is not None:
+        log.debug("Set active_low {0}: {1}".format(pin, active_low))
+        f_active_low = _open[pin].active_low
+        if active_low:
```
instead of if/else

`_write(f_active_low, int(active_low))`

You've already checked extensively that it is a boolean type!

mcampbellizo replied:
good point. changed.

vitiral commented:
I also realized a possible other approach, which is to lazily open the files only when needed. So active_low would be opened if and only if it was requested (same for other files), but then would be cached.

I think that should probably happen in a separate review though. Let me know if you want to do it (I probably won't).

mcampbellizo commented:
Once again thanks for the snappy review!

I pushed up a set of changes addressing your feedback.

As for the lazy initialization, I agree that's a good improvement, but unfortunately I don't have the time at the moment to take it on either. If I do, I'll open a PR in the future.

vitiral commented:
Do you need this on pypi?

mcampbellizo commented:
That would be great if you can publish to pypi with a minor version bump. That's our default for pulling in python dependencies.

Also thanks again for the swift review. I really enjoyed the back-and-forth with you! cheers 🍻

vitiral commented:
having issues with pypi right now, will probably be next week.

mcampbellizo commented:
No problem, thanks for everything up to this point. Let me know if there is anything I can help with.
